src/vos/vos_obj.c

```diff
@@ -266,15 +353,26 @@ vos_obj_punch(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
 
 	/* NB: punch always generate a new incarnation of the object */
 	rc = vos_obj_hold(vos_obj_cache_current(), vos_hdl2cont(coh), oid, &epr,
-			  false, DAOS_INTENT_PUNCH, true, &obj, ts_set);
+			  (flags & VOS_OF_REPLAY_PC) ? false : true,
+			  DAOS_INTENT_PUNCH, true, &obj, ts_set);
 	if (rc == 0) {
 		if (dkey) { /* key punch */
 			if (vos_ts_check_rl_conflict(ts_set, epoch))
 				read_conflict = true;
 
 			rc = key_punch(obj, epoch, pm_ver, dkey,
 				       akey_nr, akeys, flags, ts_set);
-		} else { /* object punch */
+
+			if (!read_conflict && rc > 0) {
+				/** Punch the object too */
+				punch_obj = true;
```

Comment:
When "rc > 0", it could be propagating punch to dkey only, right?

Reply:
good catch, that's a bug and may explain the extra failures.    We should only return > 0 if we need to punch the object.